### DISCUSSION TITLE
Bundling 3rd Party C Libraries

### DISCUSSION THREAD

**Discussion by bencevans** [tauri-apps/tauri#5539] [Q&A]
We've been working on an application that's dependent on 3rd party C libs. In an aim to create an application where no other frameworks/packages need to be installed on the users machine we've attempted to build these 3rd party libraries statically so they're bundled within the binary Rust produces. However we're hitting quite a few hurdles in getting some of the projects compiled statically.

Is there a way to bundle 3rd party C libraries in dynamic format with Tauri build? E.g. in the .app, .exe etc. so that a user wouldn't need to go off and install OpenCV etc. prior to running a Tauri app?

**Comment by 00hello** [comment]
@bencevans We're trying to do the same thing --get a statically compiled opencv into our application. 

Have you been able to solve this yet?

**Reply by 00hello** [reply]
We followed [these instructions](https://github.com/twistedfall/opencv-rust/issues/364#issuecomment-1486473123) in the opencv-rust crate. But we're still having some trouble getting it to work with tauri.

**Reply by bencevans** [reply]
@00hello yes, building OpenCV statically using vcpkg ... https://github.com/bencevans/camtrap-detector/blob/0d1bda3e2b907e5ddd8262d36fa2def0910ca497/.github/workflows/build.yml

**Reply by 00hello** [reply]
Thanks. And how do you setup your local development environment? 

Is it the same thing? Like, when you run `cargo tauri dev`, does it recompile OpenCV too?

**Reply by bencevans** [reply]
Pretty much. It varies a bit, sometimes I use OpenCV from Homebrew and when building a release, I install OpenCV with vcpkg, uninstall from Homebrew (just to be sure) and set the VCPKG_ROOT environment variable.

**Reply by 00hello** [reply]
I installed OpenCV with vcpkg and when I do my tauri build, I keep getting errors like

`note: /usr/bin/ld: /root/build/vcpkg/installed/x64-linux/lib/libfontconfig.a(fcfreetype.c.o): undefined reference to symbol 'FT_Get_PS_Font_Info'`

There seems to be always some library that has an undefined reference. Have you ever encountered that?

Also, why do you switch back and forth between the Homebrew version and the vcpkg? Why not just stick with vcpkg one since that's what you use in production.

**Reply by bencevans** [reply]
It looks like you're missing some links to other libraries, I'm only using a subsection of OpenCV that doesn't require fonts, it looks like you're using an OpenCV build that also needs libfontconfig linked too. Adding the links can be done with a build.rs file in the root of the project like https://github.com/bencevans/rust-opencv-yolov5/blob/main/build.rs

**Reply by 00hello** [reply]
How do you install OpenCV with vcpkg locally. Do you have a set of commands you use?

**Reply by bencevans** [reply]
Installing [`vcpkg`](https://vcpkg.io/en/getting-started.html)

```
git clone https://github.com/Microsoft/vcpkg.git

# I'm currently pinned to this release; some things were broken when I tried updating a little while ago.
git checkout 2023.02.24

./vcpkg/bootstrap-vcpkg.sh

# I set VCPKG env var now, which is needed for the Rust OpenCV crate to find the builds.
cd vcpkg && export VCPKG_ROOT="$(pwd)"
```

Building OpenCV (Static) - I'm only using the DNN module. If you are still having issues, feel free to email me (on my GitHub profile) for consulting time.

```
# macOS
./vcpkg install opencv4[dnn]:x64-osx

# Windows
./vcpkg install opencv4[dnn]:x64-windows-static-md

# Linux
./vcpkg install opencv4[dnn]:x64-linux
```

**Reply by prabhat0206** [reply]
How can I do for android?
